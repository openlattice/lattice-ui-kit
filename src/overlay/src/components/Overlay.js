/*
 * @flow
 */

import { Component } from 'react';
import type { Node } from 'react';

import PropTypes from 'prop-types';
import isFunction from 'lodash/isFunction';
import { TouchScrollable } from 'react-scrolllock';
import { CSSTransition } from 'react-transition-group';

import { OverlayInnerContainer, OverlayOuterContainer } from './styled/StyledOverlayComponents';

import Portal from '../../../portal';

type Props = {
  children ? :Node;
  isScrollable ?:boolean;
  isVisible ? :boolean;
  onClose ? :() => void;
  shouldCloseOnClick ? :boolean;
  transparent ? :boolean;
};

/*
 * Inspiration:
 * https://atlaskit.atlassian.com/packages/core/blanket
 * https://github.com/segmentio/evergreen/blob/master/src/overlay/src/Overlay.js
 */
export default class Overlay extends Component<Props> {

  static propTypes = {
    children: PropTypes.node,
    isScrollable: PropTypes.bool,
    isVisible: PropTypes.bool,
    onClose: PropTypes.func,
    shouldCloseOnClick: PropTypes.bool,
    transparent: PropTypes.bool,
  }

  static defaultProps = {
    children: undefined,
    isScrollable: false,
    isVisible: false,
    onClose: undefined,
    shouldCloseOnClick: true,
    transparent: false
  }

  close = () => {

    const { onClose } = this.props;
    if (onClose && isFunction(onClose)) {
      onClose();
    }
  }

  handleOnClick = (event :SyntheticEvent<HTMLElement>) => {

    const { shouldCloseOnClick } = this.props;
    if (event.target === event.currentTarget && shouldCloseOnClick) {
      this.close();
    }
  }

  render() {

    const {
      isVisible,
      children,
      isScrollable,
      transparent
    } = this.props;

    return (
      <CSSTransition
          in={isVisible}
          mountOnEnter
          unmountOnExit
          timeout={200}
          classNames="luk-fade">
        <Portal>
          <TouchScrollable>
            <OverlayOuterContainer transparent={transparent}>
              <OverlayInnerContainer isScrollable={isScrollable} onClick={this.handleOnClick}>
                { children }
              </OverlayInnerContainer>
            </OverlayOuterContainer>
          </TouchScrollable>
        </Portal>
      </CSSTransition>
    );
  }
}
