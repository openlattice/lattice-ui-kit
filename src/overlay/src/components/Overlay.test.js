import 'jest-styled-components';

import React from 'react';
import toJson from 'enzyme-to-json';
import { mount, shallow } from 'enzyme';

import Overlay from './Overlay';
import { nope } from '../../../utils/testing/MockUtils';

const MOCK_CHILD = (
  <span>
    hello
  </span>
);

describe('overlay', () => {

  const mVisibleOverlay = mount(
    <Overlay isVisible onClick={nope}>
      { MOCK_CHILD }
    </Overlay>
  );

  const sVisibleOverlay = shallow(
    <Overlay isVisible onClick={nope}>
      { MOCK_CHILD }
    </Overlay>
  );

  const mHiddenOverlay = mount(
    <Overlay isVisible={false} onClick={nope}>
      { MOCK_CHILD }
    </Overlay>
  );

  const sHiddenOverlay = shallow(
    <Overlay isVisible={false} onClick={nope}>
      { MOCK_CHILD }
    </Overlay>
  );

  describe('snapshots', () => {

    test('should match snapshot when isVisible="true"', () => {
      expect(toJson(mVisibleOverlay)).toMatchSnapshot();
    });

    test('should match snapshot when isVisible="false"', () => {
      expect(toJson(mHiddenOverlay)).toMatchSnapshot();
    });

  });

  describe('props', () => {

    describe('children', () => {

      test('should include all children in the render tree', () => {
        const children = (
          <div id="test-id">
            <h2>
              heading
            </h2>
            <ul className="list">
              <li>
                one
              </li>
            </ul>
            <p>
              this is a paragraph
            </p>
            <button type="button">
              clicky 1
            </button>
          </div>
        );
        const wrapper = shallow(
          <Overlay isVisible onClose={nope}>
            { children }
          </Overlay>
        );
        expect(wrapper.contains(children)).toEqual(true);
      });

    });

    describe('isVisible', () => {

      test('should show the overlay', () => {
        expect(mVisibleOverlay.children()).toHaveLength(1);
        expect(mVisibleOverlay.html()).not.toBeNull();
        expect(mVisibleOverlay.prop('isVisible')).toEqual(true);
      });

      test('should hide the overlay', () => {
        expect(mHiddenOverlay.children()).toHaveLength(0);
        expect(mHiddenOverlay.html()).toBeNull();
        expect(mHiddenOverlay.prop('isVisible')).toEqual(false);
      });

      test('should toggle visibility', () => {
        const wrapper = mount(
          <Overlay isVisible={false} onClose={nope}>
            { MOCK_CHILD }
          </Overlay>
        );
        expect(wrapper.prop('isVisible')).toEqual(false);
        wrapper.setProps({ isVisible: true });
        expect(wrapper.prop('isVisible')).toEqual(true);
        wrapper.setProps({ isVisible: false });
        expect(wrapper.prop('isVisible')).toEqual(false);
      });

    });

    describe('onClick', () => {

      test('should invoke onClick handler when clicking on the overlay', () => {
        const mockOnClick = jest.fn();
        const wrapper = mount(
          <Overlay isVisible onClick={mockOnClick}>
            { MOCK_CHILD }
          </Overlay>
        );
        wrapper.find('div').last().simulate('click');
        expect(mockOnClick).toHaveBeenCalledTimes(1);
      });

    });

  });

  describe('state', () => {

    describe('isVisible', () => {

      test('should be set to true', () => {
        const wrapper = shallow(
          <Overlay isVisible onClose={nope}>
            { MOCK_CHILD }
          </Overlay>
        );
        expect(wrapper.state().isVisible).toEqual(true);
      });

      test('should be set to false', () => {
        const wrapper = shallow(
          <Overlay isVisible={false} onClose={nope}>
            { MOCK_CHILD }
          </Overlay>
        );
        expect(wrapper.state().isVisible).toEqual(false);
      });

      test('should update when props change', () => {
        const wrapper = shallow(
          <Overlay isVisible={false} onClose={nope}>
            { MOCK_CHILD }
          </Overlay>
        );
        expect(wrapper.state().isVisible).toEqual(false);
        wrapper.setProps({ isVisible: true });
        expect(wrapper.state().isVisible).toEqual(true);
        wrapper.setProps({ isVisible: false });
        expect(wrapper.state().isVisible).toEqual(false);
      });

      test('should be set to false when closing', () => {
        const wrapper = mount(
          <Overlay isVisible>
            { MOCK_CHILD }
          </Overlay>
        );
        expect(wrapper.state().isVisible).toEqual(true);
        wrapper.find('div').last().simulate('click');
        expect(wrapper.state().isVisible).toEqual(false);
      });

    });

  });

});
