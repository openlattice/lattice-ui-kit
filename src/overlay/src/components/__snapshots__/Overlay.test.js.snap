// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overlay snapshots should match snapshot when isVisible="false" 1`] = `
<Overlay
  isScrollable={false}
  isVisible={false}
  onClose={[Function]}
  shouldCloseOnClick={true}
/>
`;

exports[`overlay snapshots should match snapshot when isVisible="true" 1`] = `
.c0 {
  background-color: rgba(8,0,25,0.3);
  bottom: 0;
  display: block;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 1000;
}

.c1 {
  display: block;
  height: 100%;
  position: absolute;
  overflow-y: visible;
  width: 100%;
}

<Overlay
  isScrollable={false}
  isVisible={true}
  onClose={[Function]}
  shouldCloseOnClick={true}
>
  <Portal>
    <Portal
      containerInfo={
        .c0 {
  background-color: rgba(8,0,25,0.3);
  bottom: 0;
  display: block;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 1000;
}

.c1 {
  display: block;
  height: 100%;
  position: absolute;
  overflow-y: visible;
  width: 100%;
}

<div
          lattice-portal-inner-container=""
        >
          <div
            class="c0"
          >
            <div
              class="c1"
            >
              <span>
                hello
              </span>
            </div>
          </div>
        </div>
      }
    >
      <OverlayOuterContainer>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "StyledOverlayComponents__OverlayOuterContainer-ol4tzw-0",
                "isStatic": false,
                "lastClassName": "c0",
                "rules": Array [
                  "background-color:",
                  "rgba(8,0,25,0.3)",
                  ";bottom:0;display:block;left:0;position:fixed;right:0;top:0;z-index:1000;",
                ],
              },
              "displayName": "OverlayOuterContainer",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "StyledOverlayComponents__OverlayOuterContainer-ol4tzw-0",
              "target": "div",
              "toString": [Function],
              "usesTheme": false,
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="c0"
          >
            <OverlayInnerContainer
              isScrollable={false}
              onClick={[Function]}
            >
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "StyledOverlayComponents__OverlayInnerContainer-ol4tzw-1",
                      "isStatic": false,
                      "lastClassName": "c1",
                      "rules": Array [
                        "display:block;height:100%;position:absolute;overflow-y:",
                        [Function],
                        ";width:100%;",
                      ],
                    },
                    "displayName": "OverlayInnerContainer",
                    "foldedComponentIds": Array [],
                    "render": [Function],
                    "styledComponentId": "StyledOverlayComponents__OverlayInnerContainer-ol4tzw-1",
                    "target": "div",
                    "toString": [Function],
                    "usesTheme": false,
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
                isScrollable={false}
                onClick={[Function]}
              >
                <div
                  className="c1"
                  onClick={[Function]}
                >
                  <span>
                    hello
                  </span>
                </div>
              </StyledComponent>
            </OverlayInnerContainer>
          </div>
        </StyledComponent>
      </OverlayOuterContainer>
      <LockToggle
        accountForScrollbars={true}
        isActive={true}
      >
        <SheetProvider
          accountForScrollbars={true}
          isActive={true}
        >
          <TouchProvider
            accountForScrollbars={true}
            isActive={true}
          >
            <ScrollLock
              accountForScrollbars={true}
              isActive={true}
            />
          </TouchProvider>
        </SheetProvider>
      </LockToggle>
    </Portal>
  </Portal>
</Overlay>
`;
